### PR TITLE
Tf-98/multiple selections contact picker

### PR DESCRIPTION
**Pull request by sherlockvn**
https://github.com/linagora/twake-on-matrix/assets/43041967/1178e463-45d2-4025-a2fd-9b86e0107d69

### CODE REVIEW

File: lib/domain/usecase/fetch_contacts_interactor.dart
Comment (hoangdat) [inline lib/domain/usecase/fetch_contacts_interactor.dart:1]: please revert this change

File: lib/widgets/avatar/avatar_with_bottom_icon_widget.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import 'package:fluffychat/pages/contacts/presentation/model/presentation_contact.dart';
+import 'package:flutter/material.dart';
+import 'package:linagora_design_flutter/avatar/round_avatar.dart';
+import 'package:linagora_design_flutter/colors/linagora_ref_colors.dart';
+
+class AvatarWithBottomIconWidget extends StatelessWidget {
+
+  final void Function() onTap;
```
Comment (hoangdat) [inline lib/widgets/avatar/avatar_with_bottom_icon_widget.dart:8]: I did not see the calling for this functions?

File: lib/widgets/avatar/avatar_with_bottom_icon_widget.dart [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      width: size,
+      height: size,
+      decoration: const BoxDecoration(
+        shape: BoxShape.circle,
+      ),
+      child: Stack(
+        alignment: Alignment.bottomRight,
+        children: [
+          RoundAvatar(text: presentationContact.displayName ?? "@"),
+          Container(
```
Comment (hoangdat) [inline lib/widgets/avatar/avatar_with_bottom_icon_widget.dart:36]: is it easy for user to touch to this icon?

File: lib/pages/new_private_chat/search_contacts_controller.dart [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  void onSearchBarChanged(String keyword) {
+    if (keyword.isNotEmpty) {
+      _debouncer.setValue(keyword);
+      searchKeyword = keyword;
+    }
+  }
+
+  void dispose() {
```
Comment (hoangdat) [inline lib/pages/new_private_chat/search_contacts_controller.dart:61]: we need to close `lookupStreamController` also?

File: lib/pages/new_group/new_group.dart [hunk 4]
```diff
@@ -15,30 +19,61 @@ class NewGroup extends StatefulWidget {
...
+  }
+
+  void listenSearchContacts() {
+    searchContactsController.lookupStreamController.stream.listen((event) {
+      Logs().d('NewGroupController::_fetchRemoteContacts() - event: $event');
+      contactStreamController.add(event);
+    });
+  }
+
+  void addContact(PresentationContact contact) {
```
Comment (hoangdat) [inline lib/pages/new_group/new_group.dart:56]: `selectContact`

File: lib/pages/new_group/new_group.dart [hunk 5]
```diff
@@ -15,30 +19,61 @@ class NewGroup extends StatefulWidget {
...
+    final newSelectedContactsMap = Map<PresentationContact, bool>.from(selectedContactsMapNotifier.value);
+    newSelectedContactsMap[contact] = true;
+    selectedContactsMapNotifier.value = newSelectedContactsMap;
+
+    haveSelectedContactsNotifier.value = selectedContactsMapNotifier.value.isNotEmpty;
+  }
+
+  void removeContact(PresentationContact contact) {
```
Comment (hoangdat) [inline lib/pages/new_group/new_group.dart:64]: unselectContact

File: lib/pages/new_group/new_group.dart [hunk 6]
```diff
@@ -15,30 +19,61 @@ class NewGroup extends StatefulWidget {
...
-    );
-    if (roomID.error == null) {
-      VRouter.of(context).toSegments(['rooms', roomID.result!, 'invite']);
-    }
+  final SearchContactsController searchContactsController = SearchContactsController();
+  final FetchContactsController fetchContactsController = FetchContactsController();
+  final StreamController<Either<Failure, GetContactsSuccess>> contactStreamController = StreamController();
+
+  final ValueNotifier<Map<PresentationContact, bool>> selectedContactsMapNotifier = ValueNotifier({});
+  final ValueNotifier<bool> haveSelectedContactsNotifier = ValueNotifier(false);
```
Comment (hoangdat) [inline lib/pages/new_group/new_group.dart:27]:
```suggestion
 ValueNotifier<bool> get haveSelectedContactsNotifier => electedContactsMapNotifier.value.isNotEmpty;
```

File: lib/pages/new_group/new_group_view.dart [hunk 7]
```diff
@@ -13,46 +17,40 @@ class NewGroupView extends StatelessWidget {
...
     return Scaffold(
-      appBar: AppBar(
-        title: Text(L10n.of(context)!.createNewGroup),
+      appBar: PreferredSize(
+        preferredSize: const Size.fromHeight(64),
+        child: SearchContactAppBar(
+          title: L10n.of(context)!.newGroupChat,
+          searchContactsController: controller.searchContactsController,
+          onCloseSearchBar: () => controller.fetchContactsController.fetchCurrentTomContacts(),
+          hintText: "Who would you like to add?",
```
Comment (hoangdat) [inline lib/pages/new_group/new_group_view.dart:26]: `l10n` please

**Comment by hoangdat** [comment]
one more remark: when you selected one more contacts, scroll should not be applied for `selected widget`. (Please compare to whatsapp)

File: lib/pages/new_group/widget/selected_participants_list.dart [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      child: SingleChildScrollView(
+        scrollDirection: Axis.horizontal,
+        child: ValueListenableBuilder<Map<PresentationContact, bool>>(
+          valueListenable: contactsNotifier,
+          builder: (context, selectedContacts, child) {
+            if (selectedContacts.isEmpty) {
+              return const SizedBox.shrink();
+            }
+            return Padding(
+              padding: const EdgeInsets.symmetric(horizontal: 8.0),
```
Comment (hoangdat) [inline lib/pages/new_group/widget/selected_participants_list.dart:36]: IMO, padding is 16

File: lib/pages/new_group/widget/selected_participants_list.dart [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+              child: Row(
+                crossAxisAlignment: CrossAxisAlignment.start,
+                children: selectedContacts.keys
+                  .map((contact) => InkWell(
+                    borderRadius: BorderRadius.circular(12.0),
+                    onTap: () => widget.newGroupController.unselectContact(contact),
+                    child: Padding(
+                      padding: const EdgeInsets.all(8.0),
```
Comment (hoangdat) [inline lib/pages/new_group/widget/selected_participants_list.dart:44]: 16 also

File: lib/pages/new_group/new_group.dart [hunk 10]
```diff
@@ -15,30 +19,61 @@ class NewGroup extends StatefulWidget {
...
+    selectedContactsMapNotifier.value = newSelectedContactsMap;
+
+    haveSelectedContactsNotifier.value = selectedContactsMapNotifier.value.isNotEmpty;
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    contactStreamController.close();
+    searchContactsController.dispose();
```
Comment (hoangdat) [inline lib/pages/new_group/new_group.dart:76]: how about `fetchContactsController`? no need to `close`?

File: lib/pages/new_group/widget/contacts_selection_list.dart [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+                                isCurrentSelected != null && isCurrentSelected
+                                  ? widget.newGroupController.selectContact(contactsList[i])
+                                  : widget.newGroupController.unselectContact(contactsList[i]);
+                              },
+                            )
+                          ],
+                        ),
+                      ),
+                    ),
+                    );
```
Comment (imGok) [inline lib/pages/new_group/widget/contacts_selection_list.dart:96]: Can you fix this indentation please :D

File: lib/pages/new_group/widget/selected_participants_list.dart [hunk 12]
```diff
@@ -0,0 +1,61 @@
+
```
Comment (imGok) [inline lib/pages/new_group/widget/selected_participants_list.dart:1]: remove line

File: lib/pages/new_private_chat/fetch_contacts_controller.dart [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  void fetchCurrentTomContacts() {
+    _fetchContactsInteractor
+      .execute()
+      .listen((event) {
+        streamController.add(event);
+      });
+  }
+
+  void dispose() {
+    streamController.close();
```
Comment (imGok) [inline lib/pages/new_private_chat/fetch_contacts_controller.dart:23]: perhaps you should end the stream interactor aswell here

File: lib/widgets/avatar/avatar_with_bottom_icon_widget.dart [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  final PresentationContact presentationContact;
+
+  final double size;
+
+  final IconData icon;
+
+  const AvatarWithBottomIconWidget({
+    super.key,
+    required this.presentationContact,
+      required this.icon,
```
Comment (imGok) [inline lib/widgets/avatar/avatar_with_bottom_icon_widget.dart:17]: weird indentation

File: lib/pages/new_private_chat/fetch_contacts_controller.dart (hunk 13 above)
Comment (sherlockvn) [reply lib/pages/new_private_chat/fetch_contacts_controller.dart:23]: _fetchContactsInteractor is a Interactor, so i can't close any stream here. _fetchContactsInteractor will be disposed when i call unbind in my ContactDI. so no worry

File: lib/widgets/avatar/avatar_with_bottom_icon_widget.dart (hunk 14 above)
Comment (sherlockvn) [reply lib/widgets/avatar/avatar_with_bottom_icon_widget.dart:17]: thanks! fix it.

File: lib/pages/new_group/widget/contacts_selection_list.dart (hunk 11 above)
Comment (sherlockvn) [reply lib/pages/new_group/widget/contacts_selection_list.dart:96]: fixed!

**Review by hoangdat — APPROVED** [review]
